### ISSUE TITLE
Get progress when generating document.

### ISSUE THREAD

**Issue by PVasiliev** [QuestPDF/QuestPDF#959]
I use QuestPDF in my backend to generate documents filled with data received from the frontend. Once the document is generated it is sent back to the frontend to be displayed.

Most of the time the document will contain a lot of images which takes some time to generate. I would like to be able to display a progress bar to the user instead of a simple spinner, so that they can gauge how much time is left.

**Comment by danirzv** [comment]
you actually may, as my experience with `QuestPDF` its too fast! probably your issue is the time consumed for transferring file to your client so you may create a process bar depend on your response file size something like this:
https://stackoverflow.com/questions/2966076/getting-file-size-in-javascript

**Comment by MarcinZiabek** [comment]
Exactly! In the described scenario, the majority of generation time is most likely spent downloading all necessary assets to the server.

When generating enormous PDF documents rich in images, most of the time is spent on image optimization to reduce output file size. I recommend using images of sizes close to the target in-document resolution or the `UseOriginalImage` API on each image.

I have put a progress API on the roadmap but it is of low priority at this moment.

**Comment by PVasiliev** [comment]
In my case I have a 40 page document. Each page has a 185x170mm image create from a byte array which in turn is created from a base64 string.

I have the following steps:
`var foo = Document.Create( ... );` Execution time: 80 ms
`var pdfDoc = foo.GeneratePdf();` Execution time: 29 seconds
`File.WriteAllBytes("test.pdf", pdfReport);` Execution time: 147 ms

As you can see, in my case it's actually the pdf generation that takes the longest time, hence my request.

**Comment by MarcinZiabek** [comment]
Could you please profile your code? The time spent during generation may likely be related to asset loading and image optimization (resizing and compression). Have you tried using the `UseOriginalImage` on each image instance?

**Comment by PVasiliev** [comment]
I'm not sure how to properly profile my code. CPU usage shows that `QuestPDF.Fluent.GenerateExtensions.GeneratePdf(QuestPDF.Infrastructure.IDocument)` took about 95.6% of all cycles.

With `UseOriginalImage` the generation takes only about 8 seconds.

**Comment by PVasiliev** [comment]
I reduced my code to a minimal working example but without the images, since I can't share them:

```
List<byte[]> imgDatas = new List<byte[]>();

var foo = Document.Merge(imgDatas.Select(imgData =>
{
    return Document.Create(container =>
    {
        container.Page(page =>
        {
            page.Size(PageSizes.A4);
            page.MarginHorizontal(2, Unit.Centimetre);
            page.MarginVertical(1, Unit.Centimetre);
            page.PageColor(Colors.White);
            page.DefaultTextStyle(x => x.FontSize(12));

            page.Content()
                .Column(col =>
                {
                    col.Item().Height(185, Unit.Millimetre).Width(170, Unit.Millimetre).Image(imgData).UseOriginalImage();
                });
        });
    });
})).UseContinuousPageNumbers();

var pdfReport = foo.GeneratePdf();

File.WriteAllBytes("test.pdf", pdfReport);
```

As I already said, with `UseOriginalImage` it takes 8 seconds, without 30 seconds.
